sycl/test/on-device/xocc/simple_tests/evaluation_tree.cpp

```diff
@@ -0,0 +1,235 @@
+//     return std::forward<ret_type>(*ret);
+//   }
+// }
+
+template <typename ret_type, typename Func, typename Var>
+[[noreturn]] inline ret_type
+visit_single_impl(Func&&, std::integer_sequence<size_t>, Var&&) {
+  // assert(false && "unreachable");
+  __builtin_unreachable();
+}
+
+template <typename ret_type, typename Func, typename Var, auto First,
+          auto... Idx>
+inline ret_type visit_single_impl(Func&& f,
+                                  std::integer_sequence<size_t, First, Idx...>,
+                                  Var&& var) {
+  if (var.index() == First)
+    return std::forward<Func>(f)(std::get<First>(var));
+  return visit_single_impl<ret_type>(std::forward<Func>(f),
+                                     std::integer_sequence<size_t, Idx...> {},
+                                     std::forward<Var>(var));
+}
+
+template <typename Func, typename Var>
+decltype(auto) visit_single(Func&& f, Var&& var) {
+  assert((!var.valueless_by_exception()));
+  using ret_type =
+      std::invoke_result_t<Func, decltype(std::get<0>(std::declval<Var>()))>;
+  return visit_single_impl<ret_type>(
+      std::forward<Func>(f),
+      typename variant_trait<
+          std::remove_cv_t<std::remove_reference_t<Var>>>::indexes {},
+      std::forward<Var>(var));
+}
+} // namespace detail
+
+/// dev_visit is std::visit implementation suitable to be used in device code.
+/// this version of visit doesn't use any function pointer but uses if series
+/// which will be turned into switch case by the optimizer.
+template <typename Func, typename Var, typename... Rest>
+auto dev_visit(Func&& f, Var&& var, Rest&&... rest) {
+  if constexpr (sizeof...(Rest) == 0)
+    return detail::visit_single(std::forward<Func>(f), std::forward<Var>(var));
+  else
+    return detail::visit_single(
+        [&](auto&& First) {
+          return dev_visit(
+              [&](auto&&... Others) {
+                std::forward<Func>(f)(
+                    std::forward<decltype(First)>(First),
+                    std::forward<decltype(Others)>(Others)...);
+              },
+              std::forward<Rest>(rest)...);
+        },
+        std::forward<Var>(var));
+}
+
+
+/// This is a relative pointer behaving mostly like T*. It doesn't manage the
+/// lifetime of the data pointed to. Also this class have shalow const semantic
+/// meaning that const rel_ptr<T> is equivalent to T* const not const T *. This
+/// pointer should behave like T* for any operation except memcpy or bitcasting
+/// the representation of the pointer. When memcpy'ed to an other place the
+/// offset between the pointer's address and the data it is pointed to will stay
+/// constant. this means that if the data is copied with the same memcpy the
+/// pointer is still pointing to the memcpy'ed data.
+/// This pointer can also be used to store pointers into files.
+template <typename T, typename IntTy = std::ptrdiff_t> class rel_ptr {
+  IntTy offset;
+
+  char *get_this() const {
```

Comment:
```suggestion
  char* get_this() const {
```
and so on to be consistent with previous code